README.md
# image_selector_formfield

This widget is a form field that allows the user to pick a image from gallery (soon from camera too), and then crop it. The widget is customizable and has all the posibilities expected on a Form Field.

This widget is using and depends on this other widgets:

 - [image_picker: ^0.6.3+4 ](https://pub.dev/packages/image_picker)
 - [image_cropper: ^1.1.0](https://pub.dev/packages/image_cropper)
 - [cached_network_image: ^2.0.1-rc](https://pub.dev/packages/cached_network_image)

 So, you need to follow the Installation Guide from each of them.

## Getting Started

To use this package, you only need to install/import it to the project and then call it in the Widget's Build method.


## Usage

This widget is design to create/update a Image like a Profile Photo from a user on any app. So, if the user has a photo already, put the image url in the property ImageURL to display it. If the user tap on the widget, the picker will allow him/her to select a image, crop it and then displaying the result.

Some visuals can be customized:
 - widget Icon,
 - widget background Color
 - widget shape: Circle or Rectangle
 - Aspect Ratio: is calculated from RatioX and RatioY; default 9/16. In circular shape is locked to 1:1
 - borderRadius:  customize size in circular shape and border radius on Rectangle corners 


## Example 

```
ImageSelectorFormField()
```

![rectangle_9x16.png](rectangle_9x16.png)


```
ImageSelectorFormField(
    cropRatioX: 16,
    cropRatioY: 9,
)
```

![rectangle_16x9.png](rectangle_16x9.png)


```
ImageSelectorFormField(
    cropStyle: CropStyle.circle,
)
```

![circle.png](circle.png)


```
ImageSelectorFormField(
   cropStyle: CropStyle.circle,
   icon: Icon(Icons.add_photo_alternate,size: 100,color: Colors.green,),
   backgroundColor: Colors.blueGrey,
   errorTextStyle: TextStyle(color: Colors.red),
   onSaved: (img) {
     print("ON SAVED EJECUTADO");
   },
   validator: (img) {
     print("validator EJECUTADO");
     return "Error Text";
   },
   cropRatioX: 9,
   cropRatioY: 16,
 )
```

![custom_circle.png](custom_circle.png)


CHANGELOG.md

## [1.0.3] - may 9, 2020.

* BUG FIX - Initial image not showing
* Added onChange callback

**Thanks [Raymond Ativie](https://github.com/RaymondAtivie) for Pr's**

## [1.0.2] - march 3, 2020.

* Replace the outdated image_picker_modern for the original image_picker. 

## [1.0.1] - November 19, 2019.

* BUG FIX - Image were not being passed to "onSave" and "Validator" callbacks;
* Set flutter version in pubspec/environment to flutter: ">= 1.10.15-pre.148 < 2.0.0"

## [1.0.0] - November 19, 2019.

* First Customizable Version.


test/image_selector_form_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:image_selector_form/image_selector_form.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


test/image_selector_formfield_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:image_selector_formfield/image_selector_formfield.dart';

void main() {
}


lib/image_selector_form.dart
library image_selector_form;

import 'package:cached_network_image/cached_network_image.dart';
import 'package:image_picker_modern/image_picker_modern.dart';
import 'package:image_cropper/image_cropper.dart';

import 'dart:io';

import 'package:flutter/material.dart';
export 'package:image_cropper/src/options.dart';

class ImageSelectorFormField extends StatefulWidget {
  ImageSelectorFormField(
      {Key key,
      this.initialImage,
      this.imageURL = "",
      this.cropRatioX,
      this.cropRatioY,
      this.boxConstraints = const BoxConstraints(maxHeight: 300, maxWidth: 300),
      this.borderRadius,
      this.cropMaxWidth = 720,
      this.cropMaxHeight = 1280,
      this.cropStyle = CropStyle.rectangle,
      this.compressQuality = 90,
      this.compressFormat = ImageCompressFormat.jpg,
      this.aspectRatioPresets = CropAspectRatioPreset.values,
      this.androidUiSettings,
      this.iosUiSettings,
      this.onSaved,
      this.validator,
      this.errorTextStyle,
      this.icon,
      this.backgroundColor = Colors.black12})
      : super(key: key);

  final File initialImage;
  final String imageURL;
  final double cropRatioX;
  final double cropRatioY;
  final BoxConstraints boxConstraints;
  final double borderRadius;
  final int cropMaxWidth;
  final int cropMaxHeight;
  final CropStyle cropStyle;
  final int compressQuality;
  final List<CropAspectRatioPreset> aspectRatioPresets;
  final ImageCompressFormat compressFormat;
  final AndroidUiSettings androidUiSettings;
  final IOSUiSettings iosUiSettings;
  final TextStyle errorTextStyle;
  final Icon icon;
  final Color backgroundColor;
  final void Function(File) onSaved;
  final String Function(File) validator;
  @override
  _ImageSelectorFormFieldState createState() => _ImageSelectorFormFieldState();
}

class _ImageSelectorFormFieldState extends State<ImageSelectorFormField> {
  File _imageFile;
  String _imageURL;
  double _aspectRatio;
  double _borderRadius;

  @override
  void initState() {
    _imageFile = widget.initialImage;
    _imageURL = widget.imageURL;
    // set default aspect ratio
    if (widget.cropRatioX != null && widget.cropRatioY != null) {
      _aspectRatio = widget.cropRatioX / widget.cropRatioY;
    } else {
      _aspectRatio = 9 / 16;
    }

    if (widget.borderRadius != null) {
      _borderRadius = widget.borderRadius;
    } else {
      if (widget.cropStyle == CropStyle.circle) {
        _borderRadius = 160;
      } else {
        _borderRadius = 15;
      }
    }
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return FormField<File>(onSaved: (_imagen) {
      if (widget.onSaved != null) widget.onSaved(_imagen);
    }, validator: (_imagen) {
      if (widget.validator != null) return widget.validator(_imagen);
      return null;
    }, builder: (state) {
      if (widget.cropStyle == CropStyle.rectangle) {
        return Column(
          children: <Widget>[
            Container(
                constraints: widget.boxConstraints,
                child: ClipRRect(
                  borderRadius: BorderRadius.circular(_borderRadius),
                  child: Container(
                    color: widget.backgroundColor,
                    child: AspectRatio(
                        aspectRatio: _aspectRatio,
                        child: _InkWidget(
                          cropStyle: widget.cropStyle,
                          imageFile: _imageFile,
                          imageURL: _imageURL,
                          androidUiSettings: widget.androidUiSettings,
                          compressFormat: widget.compressFormat,
                          aspectRatioPresets: widget.aspectRatioPresets,
                          compressQuality: widget.compressQuality,
                          cropMaxHeight: widget.cropMaxHeight,
                          cropMaxWidth: widget.cropMaxWidth,
                          cropRatioX: widget.cropRatioX,
                          cropRatioY: widget.cropRatioY,
                          iosUiSettings: widget.iosUiSettings,
                          icon: widget.icon ??
                              Icon(
                                Icons.add_a_photo,
                                size: _borderRadius / 2,
                                color: Colors.black45,
                              ),
                        )),
                  ),
                )),
            state.hasError
                ? Container(
                    padding: EdgeInsets.only(top: 5),
                    child: Text(state.errorText,
                        style: this.widget.errorTextStyle))
                : Container()
          ],
        );
      } else {
        return Column(
          children: <Widget>[
            ClipRRect(
              borderRadius: BorderRadius.circular(_borderRadius),
              child: Container(
                width: _borderRadius,
                height: _borderRadius,
                decoration: BoxDecoration(color: widget.backgroundColor),
                child: FittedBox(
                    fit: BoxFit.cover,
                    child: _InkWidget(
                      cropStyle: widget.cropStyle,
                      imageFile: _imageFile,
                      imageURL: _imageURL,
                      androidUiSettings: widget.androidUiSettings,
                      compressFormat: widget.compressFormat,
                      aspectRatioPresets: widget.aspectRatioPresets,
                      compressQuality: widget.compressQuality,
                      cropMaxHeight: widget.cropMaxHeight,
                      cropMaxWidth: widget.cropMaxWidth,
                      cropRatioX: widget.cropRatioX,
                      cropRatioY: widget.cropRatioY,
                      iosUiSettings: widget.iosUiSettings,
                      borderRadius: _borderRadius,
                      icon: widget.icon ??
                          Icon(
                            Icons.add_a_photo,
                            size: _borderRadius / 2,
                            color: Colors.black45,
                          ),
                    )),
              ),
            ),
            state.hasError
                ? Container(
                    padding: EdgeInsets.only(top: 5),
                    child: Text(state.errorText,
                        style: this.widget.errorTextStyle))
                : Container()
          ],
        );
      }
    });
  }
}

class _InkWidget extends StatefulWidget {
  _InkWidget({
    Key key,
    this.imageFile,
    this.imageURL,
    this.cropStyle = CropStyle.rectangle,
    this.cropRatioX,
    this.cropRatioY,
    this.borderRadius,
    this.cropMaxWidth,
    this.cropMaxHeight,
    this.compressQuality,
    this.aspectRatioPresets,
    this.compressFormat,
    this.androidUiSettings,
    this.iosUiSettings,
    this.icon,
  }) : super(key: key);

  final File imageFile;
  final String imageURL;
  final CropStyle cropStyle;
  final double borderRadius;
  final double cropRatioX;
  final double cropRatioY;
  final int cropMaxWidth;
  final int cropMaxHeight;
  final int compressQuality;
  final List<CropAspectRatioPreset> aspectRatioPresets;
  final ImageCompressFormat compressFormat;
  final AndroidUiSettings androidUiSettings;
  final IOSUiSettings iosUiSettings;
  final Icon icon;

  @override
  __InkWidgetState createState() => __InkWidgetState();
}

class __InkWidgetState extends State<_InkWidget> {
  File _imageFile;

  Future<File> getImage() async {
    File image = await ImagePicker.pickImage(source: ImageSource.gallery);
    if (image != null) {
      File croppedFile = await ImageCropper.cropImage(
        aspectRatio: widget.cropRatioX != null && widget.cropRatioY != null
            ? CropAspectRatio(
                ratioX: widget.cropRatioX, ratioY: widget.cropRatioY)
            : null,
        sourcePath: image.path,
        maxWidth: widget.cropMaxWidth,
        maxHeight: widget.cropMaxHeight,
        cropStyle: widget.cropStyle,
        compressQuality: widget.compressQuality,
        compressFormat: widget.compressFormat,
        aspectRatioPresets: widget.aspectRatioPresets,
        androidUiSettings: widget.androidUiSettings,
        iosUiSettings: widget.iosUiSettings,
      );
      return croppedFile;
    }
    return null;
  }

  @override
  Widget build(BuildContext context) {
    return InkWell(
      child: _imageFile == null
          ? (widget.imageURL == null || widget.imageURL == "")
              ? widget.cropStyle == CropStyle.circle
                  ? // Icon to show when there isn't image - cropstyle circle
                  Container(
                      // size is necessary for Gesture area
                      width: this.widget.borderRadius,
                      height: this.widget.borderRadius,
                      child: this.widget.icon,
                    )
                  : // Icon to show when there isn't image - cropstyle rectangle
                  this.widget.icon
              : (widget.imageURL != "" && widget.imageURL != null)
                  ? CachedNetworkImage(
                      alignment: Alignment.center,
                      imageUrl: widget.imageURL,
                      fit: BoxFit.cover,
                    )
                  : Container()
          : Image.file(_imageFile),
      onTap: () async {
        await getImage().then((imagen) {
          _imageFile = imagen;
        });
        setState(() {});
      },
    );
  }
}


lib/image_selector_formfield.dart
library image_selector_formfield;

import 'package:cached_network_image/cached_network_image.dart';
import 'package:image_cropper/image_cropper.dart';

import 'dart:io';

import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';
export 'package:image_cropper/src/options.dart';

class ImageSelectorFormField extends StatefulWidget {
  ImageSelectorFormField(
      {Key key,
      this.initialImage,
      this.imageURL = "",
      this.cropRatioX,
      this.cropRatioY,
      this.boxConstraints = const BoxConstraints(maxHeight: 300, maxWidth: 300),
      this.borderRadius,
      this.cropMaxWidth = 720,
      this.cropMaxHeight = 1280,
      this.cropStyle = CropStyle.rectangle,
      this.compressQuality = 90,
      this.compressFormat = ImageCompressFormat.jpg,
      this.aspectRatioPresets = CropAspectRatioPreset.values,
      this.androidUiSettings,
      this.iosUiSettings,
      this.onSaved,
      this.onChanged,
      this.validator,
      this.errorTextStyle,
      this.icon,
      this.backgroundColor = Colors.black12})
      : super(key: key);

  final File initialImage;
  final String imageURL;
  final double cropRatioX;
  final double cropRatioY;
  final BoxConstraints boxConstraints;
  final double borderRadius;
  final int cropMaxWidth;
  final int cropMaxHeight;
  final CropStyle cropStyle;
  final int compressQuality;
  final List<CropAspectRatioPreset> aspectRatioPresets;
  final ImageCompressFormat compressFormat;
  final AndroidUiSettings androidUiSettings;
  final IOSUiSettings iosUiSettings;
  final TextStyle errorTextStyle;
  final Icon icon;
  final Color backgroundColor;
  final void Function(File) onSaved;
  final void Function(File) onChanged;
  final String Function(File) validator;
  @override
  _ImageSelectorFormFieldState createState() => _ImageSelectorFormFieldState();
}

class _ImageSelectorFormFieldState extends State<ImageSelectorFormField> {
  File _imageFile;
  String _imageURL;
  double _aspectRatio;
  double _borderRadius;

  void _setImage(imagen) {
    if (_imageFile != imagen && widget.onChanged != null) {
      widget.onChanged(imagen);
    }
    _imageFile = imagen;
  }

  @override
  void initState() {
    _imageFile = widget.initialImage;
    _imageURL = widget.imageURL;
    // set default aspect ratio
    if (widget.cropRatioX != null && widget.cropRatioY != null) {
      _aspectRatio = widget.cropRatioX / widget.cropRatioY;
    } else {
      _aspectRatio = 9 / 16;
    }

    if (widget.borderRadius != null) {
      _borderRadius = widget.borderRadius;
    } else {
      if (widget.cropStyle == CropStyle.circle) {
        _borderRadius = 160;
      } else {
        _borderRadius = 15;
      }
    }
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return FormField<File>(onSaved: (_) {
      if (widget.onSaved != null) return widget.onSaved(_imageFile);
      return null;
    }, validator: (_) {
      if (widget.validator != null) return widget.validator(_imageFile);
      return null;
    }, builder: (state) {
      if (widget.cropStyle == CropStyle.rectangle) {
        return Column(
          children: <Widget>[
            Container(
                constraints: widget.boxConstraints,
                child: ClipRRect(
                  borderRadius: BorderRadius.circular(_borderRadius),
                  child: Container(
                    color: widget.backgroundColor,
                    child: AspectRatio(
                        aspectRatio: _aspectRatio,
                        child: _InkWidget(
                          cropStyle: widget.cropStyle,
                          imageFile: _imageFile,
                          imageURL: _imageURL,
                          androidUiSettings: widget.androidUiSettings,
                          compressFormat: widget.compressFormat,
                          aspectRatioPresets: widget.aspectRatioPresets,
                          compressQuality: widget.compressQuality,
                          cropMaxHeight: widget.cropMaxHeight,
                          cropMaxWidth: widget.cropMaxWidth,
                          cropRatioX: widget.cropRatioX,
                          cropRatioY: widget.cropRatioY,
                          iosUiSettings: widget.iosUiSettings,
                          setImage: _setImage,
                          icon: widget.icon ??
                              Icon(
                                Icons.add_a_photo,
                                size: widget.boxConstraints.biggest.width >
                                        widget.boxConstraints.biggest.height
                                    ? widget.boxConstraints.biggest.height / 2.5
                                    : widget.boxConstraints.biggest.width / 2.5,
                                color: Colors.black45,
                              ),
                        )),
                  ),
                )),
            state.hasError
                ? Container(
                    padding: EdgeInsets.only(top: 5),
                    child: Text(state.errorText,
                        style: this.widget.errorTextStyle))
                : Container()
          ],
        );
      } else {
        return Column(
          children: <Widget>[
            ClipRRect(
              borderRadius: BorderRadius.circular(_borderRadius),
              child: Container(
                width: _borderRadius,
                height: _borderRadius,
                decoration: BoxDecoration(color: widget.backgroundColor),
                child: FittedBox(
                    fit: BoxFit.cover,
                    child: _InkWidget(
                      cropStyle: widget.cropStyle,
                      imageFile: _imageFile,
                      imageURL: _imageURL,
                      androidUiSettings: widget.androidUiSettings,
                      compressFormat: widget.compressFormat,
                      aspectRatioPresets: widget.aspectRatioPresets,
                      compressQuality: widget.compressQuality,
                      cropMaxHeight: widget.cropMaxHeight,
                      cropMaxWidth: widget.cropMaxWidth,
                      cropRatioX: widget.cropRatioX,
                      cropRatioY: widget.cropRatioY,
                      iosUiSettings: widget.iosUiSettings,
                      borderRadius: _borderRadius,
                      setImage: _setImage,
                      icon: widget.icon ??
                          Icon(
                            Icons.add_a_photo,
                            size: _borderRadius / 2,
                            color: Colors.black45,
                          ),
                    )),
              ),
            ),
            state.hasError
                ? Container(
                    padding: EdgeInsets.only(top: 5),
                    child: Text(state.errorText,
                        style: this.widget.errorTextStyle))
                : Container()
          ],
        );
      }
    });
  }
}

class _InkWidget extends StatefulWidget {
  _InkWidget(
      {Key key,
      this.imageFile,
      this.imageURL,
      this.cropStyle = CropStyle.rectangle,
      this.cropRatioX,
      this.cropRatioY,
      this.borderRadius,
      this.cropMaxWidth,
      this.cropMaxHeight,
      this.compressQuality,
      this.aspectRatioPresets,
      this.compressFormat,
      this.androidUiSettings,
      this.iosUiSettings,
      this.icon,
      this.setImage})
      : super(key: key);

  final File imageFile;
  final String imageURL;
  final CropStyle cropStyle;
  final double borderRadius;
  final double cropRatioX;
  final double cropRatioY;
  final int cropMaxWidth;
  final int cropMaxHeight;
  final int compressQuality;
  final List<CropAspectRatioPreset> aspectRatioPresets;
  final ImageCompressFormat compressFormat;
  final AndroidUiSettings androidUiSettings;
  final IOSUiSettings iosUiSettings;
  final Icon icon;
  final Function(File) setImage;

  @override
  __InkWidgetState createState() => __InkWidgetState();
}

class __InkWidgetState extends State<_InkWidget> {
  File _imageFile;

  @override
  void initState() {
    super.initState();
    if (widget.imageFile != null) {
      _imageFile = widget.imageFile;
    }
  }

  Future<File> getImage() async {
    File image = await ImagePicker.pickImage(source: ImageSource.gallery);
    if (image != null) {
      File croppedFile = await ImageCropper.cropImage(
        aspectRatio: widget.cropStyle == CropStyle.circle
            ? CropAspectRatio(ratioX: 1.0, ratioY: 1.0)
            : widget.cropRatioX != null && widget.cropRatioY != null
                ? CropAspectRatio(
                    ratioX: widget.cropRatioX, ratioY: widget.cropRatioY)
                : null,
        sourcePath: image.path,
        maxWidth: widget.cropMaxWidth,
        maxHeight: widget.cropMaxHeight,
        cropStyle: widget.cropStyle,
        compressQuality: widget.compressQuality,
        compressFormat: widget.compressFormat,
        aspectRatioPresets: widget.aspectRatioPresets,
        androidUiSettings: widget.androidUiSettings,
        iosUiSettings: widget.iosUiSettings,
      );
      return croppedFile;
    }
    return null;
  }

  @override
  Widget build(BuildContext context) {
    return InkWell(
      child: _imageFile == null
          ? (widget.imageURL == null || widget.imageURL == "")
              ? widget.cropStyle == CropStyle.circle
                  ? // Icon to show when there isn't image - cropstyle circle
                  Container(
                      // size is necessary for Gesture area
                      width: this.widget.borderRadius,
                      height: this.widget.borderRadius,
                      child: this.widget.icon,
                    )
                  : // Icon to show when there isn't image - cropstyle rectangle
                  this.widget.icon
              : (widget.imageURL != "" && widget.imageURL != null)
                  ? CachedNetworkImage(
                      alignment: Alignment.center,
                      imageUrl: widget.imageURL,
                      fit: BoxFit.cover,
                    )
                  : Container()
          : Image.file(_imageFile),
      onTap: () async {
        await getImage().then((imagen) {
          _imageFile = imagen;
          widget.setImage(imagen);
        });
        setState(() {});
      },
    );
  }
}
